adapters/madvertise/madvertise.go

```diff
@@ -0,0 +1,170 @@
+package madvertise
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpointTemplate template.Template
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpointTemplate: *template,
+	}
+
+	return bidder, nil
+}
+
+func getHeaders(request *openrtb2.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return &headers
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	placementID := ""
+	for _, imp := range request.Imp {
+		madvertiseExt, err := getImpressionExt(imp)
+		if err != nil {
+			return nil, []error{err}
+		}
+		if madvertiseExt.PlacementId != "" {
+			if placementID == "" {
+				placementID = madvertiseExt.PlacementId
+			} else if placementID != madvertiseExt.PlacementId {
```

Comment:
Can you please add a JSON test to verify this condition as well?

Reply:
Above there is 
`madvertiseExt, err := getImpressionExt(imp)
		if err != nil {
			return nil, []error{err}
		}`

with

`func getImpressionExt(imp openrtb2.Imp) (*openrtb_ext.ExtImpMadvertise, error) {
	var bidderExt adapters.ExtImpBidder
	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
		return nil, &errortypes.BadInput{
			Message: "Bidder extension not provided or can't be unmarshalled",
		}
	}

	var madvertiseExt openrtb_ext.ExtImpMadvertise
	if err := json.Unmarshal(bidderExt.Bidder, &madvertiseExt); err != nil {
		return nil, &errortypes.BadInput{
			Message: "Error while unmarshaling bidder extension",
		}
	}

	return &madvertiseExt, nil
}`

Reply:
Not sure where so you mean the test exists but what I was recommending was to add another JSON test in `adapters/madvertise/madvertisetest/supplemental/` where the request has multiple imps but with different `zoneId` values. This should trigger the `if` condition above and return an error.

Reply:
Ok Sorry.
I push [master 33d69138] Add JSON test unique-zone-id.json in order to add new JSON test.